### PR TITLE
feat: Image cropping

### PR DESCRIPTION
**Pull request by wbm1113**
Issue #4236.  Note: this is my first pull request (ever).  Beware?

This PR adds cropping capability.  When selecting a single image, a new cropping button appears.  Click the button to enter cropping mode; click it again to exit cropping mode.  ~~It currently works in the 4 main cardinal directions.~~  Now works with corners too.  It handles arbitrary scale, rotation, and inversion.

Example:

image

Definitely needs to be done before taking this out of draft status:
- ~~Still need support for cropping from corners (only N/S/E/W works at the moment)~~
- ~~Still need to write tests~~
- ~~Still need to test on:~~
    - ~~Firefox~~
    - ~~Edge~~
    - ~~Safari~~
    - ~~Mobile~~
 
Remarks:
- My crop icon doesn't look quite right, but I'm not sure why
- I was utterly confused by the return value of `actionCropImage.perform()`.  I'm pretty sure it's not set up correctly.  ~~I was having transient issues with undo/redo that I assume are related to this.~~
- Issue #4236 set forth the desired implementation for crop transform handles.  I fell short on this and opted to signify "cropping mode" by changing the transform handle colors.  The desired implementation would have required significant restructuring of how the transform handles are built.  We would need to add support for:
    - Rectangular transform handles (easy)
    - L-shaped transform handles (not so easy)
    - Rotating transform handles
    - Hit detection on these new shapes.  From what I'm seeing, the rendering computes the dimensions of the transform handles.  Then, the hit detection re-computes them.  First step toward this would probably be refactoring that so there is a single source of truth for the transform handle values, if that is possible.
- I hard-coded the updated crop handle color (#228be6) into `renderScene.ts`.  I'd welcome a better idea on where to put this.
- I added a tooltip to the crop button that just says "Crop," which will need translation.
- ~~Any time you mouse down in a way that doesn't crop, it cancels the crop state.  E.g. if you crop, then move the image, moving the image cancels the crop.  This might be annoying.  Same with rotation.~~
- ~~Mouse up events intermittently drop the image selection and you have to re-select the image, then re-toggle crop mode to keep cropping.~~
- The blue highlight on the cropping mode button sometimes disappears in Safari
- When cropping inverted images, the cropping behavior is also inverted (image below)

image

### CODE REVIEW

**Comment by zsviczian** [comment]
This is a very useful feature!

I would recommend/prefer that we solve cropping by binding the image to a Rectangle container. We could use the same SHIFT+resize to resize the cropped area while normal resize would resize the image.

Binding the image to a rectangle for cropping would open the way to later implementing binding to a diamond and an ellipse, maybe to a linear element closed path at some point in the future.

see: #4566

**Comment by wbm1113** [comment]
I've only worked in this project for like a week, so take this with a grain of salt.

I think you're spot in everything you said.  I.e., your suggestion is an outright better implementation.  But, my intuition is to make a separate PR for it.  The binding system looks like it's only configured to allow arrows to be binding elements:

```
export const isBindingElementType = (
  elementType: AppState["activeTool"]["type"],
): boolean => {
  return elementType === "arrow";
};
```

More generically, the binding functions are heavily dependent on dealing with an `ExcalidrawLinearElement`.  So, images would need a points property and start/end binding.  You'd have to either copy and paste those properties from the linear element to the image element, or restructure how the elements inherit from one another.  Or, make a new set of binding functions that works on images.  In any case, this seems like a significant refactor.  If I'm wrong there, please correct me.

I'd hate to see this otherwise serviceable PR, which contains a useful feature which has been requested since 2021, be gated behind that.  The cropping logic is reasonably encapsulated and could be easily transplanted to a shape-binding-based system once that is ready.

Summarily, https://www.youtube.com/watch?v=SETnK2ny1R0

File: src/renderer/renderScene.ts [hunk 1]
```diff
@@ -848,15 +860,23 @@ const renderTransformHandles = (
...
         context.strokeStyle = renderConfig.selectionColor;
       }
+
       if (key === "rotation") {
+        context.fillStyle = oc.white;
         fillCircle(context, x + width / 2, y + height / 2, width / 2);
         // prefer round corners if roundRect API is available
       } else if (context.roundRect) {
+        if (cropModeEnabled) {
+          context.fillStyle = "#228be6";
```
Comment (DanielJGeiger) [inline src/renderer/renderScene.ts:870]: Did you mean to hardcode this?  (Line 878 also.)

Comment (wbm1113) [reply src/renderer/renderScene.ts:870]: Yes and no.  I didn't know where else to put it.  The color fits both light and dark themes, so I considered it workable for now (whether that's right or wrong, I don't know).

**Comment by DanielJGeiger** [comment]
> First round of tests failed on this because I forgot to lint. My bad there. This should be good to re-test after my commit from 04/27.

Could you push another commit?  (I tried re-approving the tests recently.)

**Comment by wbm1113** [comment]
I made a small change to a comment, re-ran the lint/test commands, then committed.  Not sure how/why, but the Vercel build failed.

**Comment by DanielJGeiger** [comment]
@wbm1113 I'm not sure what went wrong with Vercel either (not specific to this PR?).  But I just approved the GH test/lint checks again for you.

**Comment by dwelle** [comment]
For some reason we've having issues with Vercel builds lately, and most PRs fail on an error similar to this:

image

The solution is to open the vercel build and redeploy manually.

**Comment by wbm1113** [comment]
Added a couple commits to resolve merge conflicts

**Comment by dwelle** [comment]
@wbm1113 hi. Apologies for dropping the ball on this. Are you still available to work on this? If so, we'll make a proper review and point what needs to be changed/ developed further. 🙏

**Comment by wbm1113** [comment]
@dwelle Yes.  While I'm sure you are more than capable of doing the review on your own, the 'remarks' section in my original comment has a few bullet points where I felt a review was very much needed.

**Comment by mindriven** [comment]
would love to see this one!

File: src/actions/actionCropImage.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    const nextAppState = {
+      ...appState,
+      croppingModeEnabled,
+    };
+
+    return {
+      elements,
+      appState: {
+        ...nextAppState,
```
Comment (DanielJGeiger) [inline src/actions/actionCropImage.tsx:33]:
```suggestion
        ...appState,
        croppingModeEnabled,
```

File: src/actions/actionCropImage.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      getNonDeletedElements(elements),
+      appState,
+    );
+    if (selectedElements.length === 1 && selectedElements[0].type === "image") {
+      croppingModeEnabled = !croppingModeEnabled;
+    } else {
+      croppingModeEnabled = false;
+    }
+
+    const nextAppState = {
```
Comment (DanielJGeiger) [inline src/actions/actionCropImage.tsx:25]: I think you could drop these four lines with the change I suggested on line 33.

File: src/actions/actionCropImage.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      ...appState,
+      croppingModeEnabled,
+    };
+
+    return {
+      elements,
+      appState: {
+        ...nextAppState,
+        activeTool: updateActiveTool(appState, { type: "selection" }),
+        multiElement: null,
```
Comment (DanielJGeiger) [inline src/actions/actionCropImage.tsx:35]:
```suggestion
```
Do you need this line?  The type of `multiElement` involves `ExcalidrawLinearElement`.

File: src/actions/actionCropImage.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    const nextAppState = {
+      ...appState,
+      croppingModeEnabled,
+    };
+
+    return {
+      elements,
+      appState: {
+        ...nextAppState,
+        activeTool: updateActiveTool(appState, { type: "selection" }),
```
Comment (DanielJGeiger) [inline src/actions/actionCropImage.tsx:34]: This line looks redundant.  Can you access `actionCropImage` without already having `activeTool.type === "selection"`?

**Review by DanielJGeiger — COMMENTED** [review]
@wbm1113 I've attempted a reply to your remark about the return value of `actionCropImage.perform()`.  @dwelle could probably give you a better answer though.

**Comment by chrisdoherty4** [comment]
> Click the button to enter cropping mode; click it again to exit cropping mode

Great PR, great feature. I wondered if it makes sense to allow the user to hit `ESC` also to exit crop mode. Perhaps its just me, but hitting escape feels natural and more convenient than mouse clicking a button.

**Comment by Neilblaze** [comment]
> if it makes sense to allow the user to hit `ESC` also to exit crop mode. 

+1 from my end. Rest, this feature needs to be merged asap.

**Review by jiweiyuan — COMMENTED** [review]
Pretty good PR

**Comment by IntelligentQuantum** [comment]
Status?

**Comment by dwelle** [comment]
Work will restart in the matter of weeks.

**Comment by ryan-di** [comment]
Hey @wbm1113, sorry again for dropping the ball. We've finally found time to get to this long awaited feature. Given how much the codebase has changed, we think it makes more sense to open a separate PR to start the work from scratch. Thanks very much for spearheading the work and much appreciated for the review by @DanielJGeiger as well ❤️

**Pull request closed without merging**